sql/core/src/test/java/test/org/apache/spark/sql/JavaBeanWithMapSuite.java

```diff
@@ -0,0 +1,257 @@
+
+        Record(int id, Map<String, Interval> intervals, Map<String, Integer> values) {
+            this.id = id;
+            this.intervals = intervals;
+            this.values = values;
+        }
+
+        public int getId() {
+            return id;
+        }
+
+        public void setId(int id) {
+            this.id = id;
+        }
+
+        public Map<String, Interval> getIntervals() {
+            return intervals;
+        }
+
+        public void setIntervals(Map<String, Interval> intervals) {
+            this.intervals = intervals;
+        }
+
+        public Map<String, Integer> getValues() {
+            return values;
+        }
+
+        public void setValues(Map<String, Integer> values) {
+            this.values = values;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (!(obj instanceof Record)) return false;
+            Record other = (Record) obj;
+            return
+                    (other.id == this.id) &&
+                    Util.equals(other.intervals, this.intervals) &&
+                    Util.equals(other.values, this.values);
+        }
+
+        @Override
+        public String toString() {
+            return String.format("{ id: %d, intervals: %s, values: %s }", id, intervals, values);
+        }
+    }
+
+    public static class Interval {
+
+        private long startTime;
+        private long endTime;
+
+        public Interval() { }
+
+        Interval(long startTime, long endTime) {
+            this.startTime = startTime;
+            this.endTime = endTime;
+        }
+
+        public long getStartTime() {
+            return startTime;
+        }
+
+        public void setStartTime(long startTime) {
+            this.startTime = startTime;
+        }
+
+        public long getEndTime() {
+            return endTime;
+        }
+
+        public void setEndTime(long endTime) {
+            this.endTime = endTime;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (!(obj instanceof Interval)) return false;
+            Interval other = (Interval) obj;
+            return
+                    (other.startTime == this.startTime) &&
+                    (other.endTime == this.endTime);
+        }
+
+        @Override
+        public String toString() {
+            return String.format("[%d,%d]", startTime, endTime);
+        }
+    }
+
+    private static class Util {
+
+        private static <E> boolean equals(Collection<E> as, Collection<E> bs) {
+            if (as.size() != bs.size()) return false;
+            Iterator<E> ai = as.iterator();
+            Iterator<E> bi = bs.iterator();
+            while (ai.hasNext() && bi.hasNext()) {
+                if (!ai.next().equals(bi.next())) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        private static <K, V> boolean equals(Map<K, V> aMap, Map<K, V> bMap) {
```

Comment:
hmm, do we need this specific `equals`? Can't we use `Map.equals`?

Reply:
Yes, sure, that's absolutely redundand code...